util/topsql/reporter/reporter.go

```diff
@@ -471,7 +512,12 @@ func (tsr *RemoteTopSQLReporter) reportWorker() {
 	defer util.Recover("top-sql", "reportWorker", nil, false)
 
 	for {
+		tsr.removeDownDataSinks()
+		tsr.keepDataSinksFirst(10)
+
 		select {
+		case dataSink := <-tsr.dataSinkRegCh:
+			tsr.dataSinks = append(tsr.dataSinks, dataSink)
```

Comment:
I guess it's fine to introduce a mutex for registering and reading the data sink, as it is a very low frequent operation (and reading / modifying the data sink list is also very quick). Then you will find the design much simplified. There is no need for everything to be channel based. This also makes it possible to unregister the data sink quite easily.

Reply:
It feels more comfortable for me to think of these as actor models. I am not good at figuring out the control flow of mutex models.

Reply:
BTW, it's my fault to cleaning up `dataSinks` at `RemoteTopSQLReporter.Close`...